File: SBiswas/Problems/Milestone1/Problem15.cpp
/* 110. Balanced Binary Tree*/

class TreeNode {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};


class Solution {
public:
    bool isBalanced(TreeNode* root) {
        
        if(getHeight(root)==-1)
        {
            return false;
        }

        return true;
    }

    int getHeight(TreeNode* root)
    {   
        if(!root)
        {
            return 0;
        }

        int left_subtree_height = getHeight(root->left); // get left subtree height
        int right_subtree_height = getHeight(root->right); // get right subtree height
        
        if(left_subtree_height == -1 || right_subtree_height == -1) // if either of the heights are -1 then return -1
        {
            return -1;
        }

        if(abs(left_subtree_height-right_subtree_height)>1) // check if the height diff between left and right is greater than 1.
        {
            return -1;
        }

        return max(left_subtree_height,right_subtree_height)+1; // return the height of the current subtree with root as root_node.
        
    }
};

File: VishnuThakral/Milestone1/Problem15.py
# Milestone1 Problem 15 - Leetcode 110(Balanced Binary Tree: https://leetcode.com/problems/balanced-binary-tree/description/)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def height(node):
            if not node:
                return -1
            else:
                return 1 + max(height(node.left), height(node.right))

        if not root:
            return True
        else:
            return abs(height(root.left) - height(root.right)) < 2 and self.isBalanced(root.left) and self.isBalanced(
                root.right)